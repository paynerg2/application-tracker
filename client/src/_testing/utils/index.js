import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../../_reducers';
import { middlewares } from '../../app/store';

export const testStore = (initialState) => {
    const createStoreWithMiddleware = applyMiddleware(...middlewares)(createStore);
    return createStoreWithMiddleware(rootReducer, initialState);
};
