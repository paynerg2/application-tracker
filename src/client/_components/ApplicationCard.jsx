import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import { applicationActions, interviewActions } from '../_actions';
import { ProgressBar, InterviewList } from '../_components';
import { CardListItem, Dropdown, Button } from './Card.styles';

const RightColumn = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
    align-items: center;
`;

const LeftColumn = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
`;

const Bottom = styled.div`
    display: flex;
    width: 100%;
    flex-direction: column;
    justify-content: space-between;
`;

const Title = styled.div`
    font-weight: bold;
    border-bottom: solid 1px rgba(0, 0, 0, 0.0975);
`;

const Location = styled.div`
    font-style: italic;
`;

class ApplicationCard extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isSelected: false
        };

        this.handleDelete = this.handleDelete.bind(this);
        this.handleResponseUpdate = this.handleResponseUpdate.bind(this);
    }

    handleDelete(id) {
        // Delete the selected application, and any associated interviews.
        const { dispatch, interviews } = this.props;
        dispatch(applicationActions.delete(this.props.application._id));
        interviews.forEach(interview =>
            dispatch(interviewActions.delete(interview._id))
        );
    }

    handleResponseUpdate(e) {
        const { dispatch } = this.props;
        const { id } = this.props.application;
        const selectedApplication = this.props.applicationList.find(
            app => app.id === id
        );

        if (selectedApplication) {
            dispatch(
                applicationActions.update(id, {
                    ...selectedApplication,
                    response: e.target.value
                })
            );
        }

        if (e.target.value === 'Interview') {
            this.props.history.push(`applications/${id}/interviews`);
        }
    }

    formatDate(date) {
        if (date) {
            const d = new Date(date);
            return d.toLocaleDateString('en-US');
        }
        return '';
    }

    renderCard = application => {
        const {
            _id,
            jobTitle,
            company,
            location,
            mainSkill,
            dateApplicationSent,
            response,
            requiredSkillsMet,
            requiredSkillsTotal,
            additionalSkillsMet,
            additionalSkillsTotal
        } = application;
        const { interviews } = this.props;
        return (
            <React.Fragment>
                <LeftColumn>
                    <div>
                        <Title>{jobTitle}</Title>
                        <div>{company}</div>
                    </div>
                    <Location>{location}</Location>
                    {!(interviews.length > 0) && (
                        <Dropdown
                            value={response}
                            onChange={e => this.handleResponseUpdate(e)}
                        >
                            <option value="No Response">No Response</option>
                            <option value="Rejected">Rejected</option>
                            <option value="Interview">Add Interview</option>
                        </Dropdown>
                    )}
                    <Bottom>
                        <div>
                            {response === 'Interview' && (
                                <React.Fragment>
                                    <InterviewList interviews={interviews} />
                                    <Button
                                        onClick={() =>
                                            this.props.history.push(
                                                `/applications/${_id}/interviews`
                                            )
                                        }
                                    >
                                        +
                                    </Button>
                                </React.Fragment>
                            )}
                        </div>
                        <div>
                            <ProgressBar
                                fill={
                                    requiredSkillsTotal > 0
                                        ? (requiredSkillsMet /
                                              requiredSkillsTotal) *
                                          100
                                        : 0
                                }
                            />
                            <ProgressBar
                                fill={
                                    additionalSkillsTotal > 0
                                        ? (additionalSkillsMet /
                                              additionalSkillsTotal) *
                                          100
                                        : 0
                                }
                            />
                        </div>
                    </Bottom>
                </LeftColumn>
                <RightColumn>
                    {this.state.isSelected && (
                        <div>
                            <Link to={`/applications/${_id}`}>
                                <Button>edit</Button>
                            </Link>
                            <Button onClick={this.handleDelete}>Delete</Button>
                        </div>
                    )}
                    <div>{mainSkill}</div>
                    <div>{this.formatDate(dateApplicationSent)}</div>
                </RightColumn>
            </React.Fragment>
        );
    };

    render() {
        return (
            <CardListItem
                onMouseEnter={() => this.setState({ isSelected: true })}
                onMouseLeave={() => this.setState({ isSelected: false })}
            >
                {this.renderCard(this.props.application)}
            </CardListItem>
        );
    }
}

function mapStateToProps(state) {
    const { applicationList } = state.applications;
    return {
        applicationList
    };
}

const connectedApplicationCard = connect(mapStateToProps)(ApplicationCard);
export { connectedApplicationCard as ApplicationCard };
