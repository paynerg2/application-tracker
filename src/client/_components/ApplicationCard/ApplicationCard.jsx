import React, { Component } from 'react';
import { PropTypes } from 'prop-types';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import styled from 'styled-components';

import { applicationActions, interviewActions } from '../../_actions';
import { ProgressBar } from '../ProgressBar';
import { InterviewList } from '../InterviewList';
import { CardListItem, Dropdown, Button } from '../Card.styles';
import { ReactComponent as EditIcon } from '../../_assets/icons8-edit-50.svg';
import { ReactComponent as DeleteIcon } from '../../_assets/icons8-trash-50.svg';
import { ReactComponent as DownChevron } from '../../_assets/down-chevron.svg';
import { iconSelector } from '../../_helpers/iconSelector';

const RightColumn = styled.div`
    display: grid;
    grid-template-rows: 30% 30% 30%;
    justify-items: center;
    align-items: center;
    align-content: space-between;
`;
RightColumn.displayName = 'RightColumn';

const LeftColumn = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
`;
LeftColumn.displayName = 'LeftColumn';

const Bottom = styled.div`
    display: flex;
    width: 100%;
    flex-direction: column;
    justify-content: space-between;
`;
Bottom.displayName = 'Bottom';

const Title = styled.div`
    font-weight: bold;
    border-bottom: solid 1px rgba(0, 0, 0, 0.0975);
`;
Title.displayName = 'Title';

const Location = styled.div`
    font-style: italic;
`;
Location.displayName = 'Location';

const UpdateButton = styled(Button)`
    box-shadow: 1px 2px 6px 1px darkgrey;
`;
UpdateButton.displayName = 'UpdateButton';

class ApplicationCard extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isSelected: false,
            showEditButtons: false
        };

        this.handleDelete = this.handleDelete.bind(this);
        this.handleResponseUpdate = this.handleResponseUpdate.bind(this);
    }

    handleDelete(id) {
        // Delete the selected application, and any associated interviews.
        const { dispatch, interviews } = this.props;
        dispatch(applicationActions.delete(this.props.application._id));
        interviews.forEach(interview =>
            dispatch(interviewActions.delete(interview._id))
        );
    }

    handleResponseUpdate(e) {
        const { dispatch } = this.props;
        const { id } = this.props.application;
        const selectedApplication = this.props.applicationList.find(
            app => app.id === id
        );

        if (selectedApplication) {
            dispatch(
                applicationActions.update(id, {
                    ...selectedApplication,
                    response: e.target.value
                })
            );
        }

        if (e.target.value === 'Interview') {
            this.props.history.push(`applications/${id}/interviews`);
        }
    }

    formatDate(date) {
        if (date) {
            const d = new Date(date);
            return d.toLocaleDateString('en-US');
        }
        return '';
    }

    renderCard = application => {
        const {
            _id,
            jobTitle,
            company,
            location,
            mainSkill,
            dateApplicationSent,
            response,
            requiredSkillsMet,
            requiredSkillsTotal,
            additionalSkillsMet,
            additionalSkillsTotal
        } = application;
        const { interviews } = this.props;
        return (
            <React.Fragment>
                <LeftColumn>
                    <div>
                        <Title>{jobTitle}</Title>
                        <div>{company}</div>
                    </div>
                    <Location>{location}</Location>

                    <Bottom>
                        <div>
                            {interviews.length === 0 && (
                                <Dropdown
                                    value={response}
                                    onChange={e => this.handleResponseUpdate(e)}
                                >
                                    <option value="No Response">
                                        No Response
                                    </option>
                                    <option value="Rejected">Rejected</option>
                                    <option value="Interview">
                                        Add Interview
                                    </option>
                                </Dropdown>
                            )}
                            {response === 'Interview' && (
                                <React.Fragment>
                                    <InterviewList interviews={interviews} />
                                    <Button
                                        onClick={() =>
                                            this.props.history.push(
                                                `/applications/${_id}/interviews`
                                            )
                                        }
                                    >
                                        +
                                    </Button>
                                </React.Fragment>
                            )}
                        </div>
                        <div>
                            <ProgressBar
                                fill={
                                    requiredSkillsTotal > 0
                                        ? (requiredSkillsMet /
                                              requiredSkillsTotal) *
                                          100
                                        : 0
                                }
                            />
                            <ProgressBar
                                fill={
                                    additionalSkillsTotal > 0
                                        ? (additionalSkillsMet /
                                              additionalSkillsTotal) *
                                          100
                                        : 0
                                }
                            />
                        </div>
                    </Bottom>
                </LeftColumn>
                <RightColumn>
                    <div>
                        <div
                            onClick={() =>
                                this.setState({
                                    showEditButtons: !this.state.showEditButtons
                                })
                            }
                        >
                            <DownChevron />
                        </div>
                        {this.state.showEditButtons && (
                            <div>
                                <Link to={`/applications/${_id}`}>
                                    <UpdateButton>
                                        <EditIcon />
                                    </UpdateButton>
                                </Link>
                                <UpdateButton onClick={this.handleDelete}>
                                    <DeleteIcon />
                                </UpdateButton>
                            </div>
                        )}
                    </div>
                    <div>{iconSelector(mainSkill)}</div>
                    <div>{this.formatDate(dateApplicationSent)}</div>
                </RightColumn>
            </React.Fragment>
        );
    };

    render() {
        return (
            <CardListItem
                onMouseEnter={() => this.setState({ isSelected: true })}
                onMouseLeave={() => this.setState({ isSelected: false })}
            >
                {this.renderCard(this.props.application)}
            </CardListItem>
        );
    }
}

function mapStateToProps(state) {
    const { applicationList } = state.applications;
    return {
        applicationList
    };
}

ApplicationCard.propTypes = {
    applicationList: PropTypes.array,
    application: PropTypes.object,
    interviews: PropTypes.array,
    history: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired
};

const connectedApplicationCard = connect(mapStateToProps)(ApplicationCard);
export { connectedApplicationCard as ApplicationCard };
