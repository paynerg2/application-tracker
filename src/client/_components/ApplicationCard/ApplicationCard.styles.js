import styled from 'styled-components';

import { Button } from '../Card.styles';

export const RightColumn = styled.div`
    display: grid;
    grid-template-rows: 30% 30% 30%;
    justify-items: center;
    align-items: center;
    align-content: space-between;
`;
RightColumn.displayName = 'RightColumn';

export const LeftColumn = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: space-between;
`;
LeftColumn.displayName = 'LeftColumn';

export const Bottom = styled.div`
    display: flex;
    width: 100%;
    flex-direction: column;
    justify-content: space-between;
`;
Bottom.displayName = 'Bottom';

export const Title = styled.div`
    font-weight: bold;
    border-bottom: solid 1px rgba(0, 0, 0, 0.0975);
`;
Title.displayName = 'Title';

export const Location = styled.div`
    font-style: italic;
`;
Location.displayName = 'Location';

export const UpdateButton = styled(Button)`
    box-shadow: 1px 2px 6px 1px darkgrey;
`;
UpdateButton.displayName = 'UpdateButton';
