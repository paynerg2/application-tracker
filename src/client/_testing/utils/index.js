import checkPropTypes from 'check-prop-types';
import { applyMiddleware, createStore } from 'redux';
import rootReducer from '../../_reducers';

export const checkProps = (component, expectedProps) => {
    const propsError = checkPropTypes(
        component.propTypes,
        expectedProps,
        'props',
        component.name
    );
    return propsError;
};
