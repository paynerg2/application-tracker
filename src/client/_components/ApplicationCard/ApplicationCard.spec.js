import React from 'react';
import { shallow } from 'enzyme';

import { ApplicationCard } from './ApplicationCard';
import { testStore } from '../../_testing/utils';

const setUp = (props = {}, initialState = {}) => {
    const store = testStore(initialState);
    const component = shallow(
        <ApplicationCard store={store} {...props} />
    ).dive();
    return component;
};

describe('Application Card Component', () => {
    let mockFunction, wrapper;
    beforeEach(() => {
        mockFunction = jest.fn();
        const props = {
            dispatch: mockFunction,
            application: {
                _id: 'test',
                jobTitle: 'test',
                company: 'test',
                location: 'test',
                mainSkill: 'test',
                dateApplicationSent: '5/5/55',
                response: 'No Response',
                requiredSkillsMet: 1,
                requiredSkillsTotal: 10,
                additionalSkillsMet: 1,
                additionalSkillsTotal: 10
            },
            interviews: []
        };
        wrapper = setUp(props);
    });

    it('renders without crashing', () => {
        expect(wrapper).toBeDefined();
    });

    describe('when the component is selected', () => {
        beforeEach(() => {
            const cardListItemWrapper = wrapper.find('CardListItem');
            cardListItemWrapper.simulate('mouseEnter');
        });

        it('shows additional buttons in the right column', () => {
            const rightColumnWrapper = wrapper.find('RightColumn');
            expect(rightColumnWrapper).toMatchSnapshot();
        });
    });

    describe('when the component is not selected', () => {
        it('does not show additional buttons in the right column', () => {
            const rightColumnWrapper = wrapper.find('RightColumn');
            expect(rightColumnWrapper).toMatchSnapshot();
        });

        it('returns to unselected state on mouse leave event', () => {
            const cardListItemWrapper = wrapper.find('CardListItem');
            cardListItemWrapper.simulate('mouseEnter');
            cardListItemWrapper.simulate('mouseLeave');
            const rightColumnWrapper = wrapper.find('RightColumn');
            expect(rightColumnWrapper).toMatchSnapshot();
        });
    });

    describe('when there are no associated interviews', () => {
        it('shows a dropdown for application response', () => {
            const leftColumnWrapper = wrapper.find('LeftColumn');
            expect(leftColumnWrapper).toMatchSnapshot();
        });

        it('does not show an interview list or button to add interviews', () => {
            const bottomWrapper = wrapper.find('Bottom');
            expect(bottomWrapper).toMatchSnapshot();
        });
    });

    describe('when there are associated interviews', () => {
        beforeEach(() => {
            const props = {
                dispatch: mockFunction,
                application: {
                    _id: 'test',
                    jobTitle: 'test',
                    company: 'test',
                    location: 'test',
                    mainSkill: 'test',
                    dateApplicationSent: '5/5/55',
                    response: 'Interview',
                    requiredSkillsMet: 1,
                    requiredSkillsTotal: 10,
                    additionalSkillsMet: 1,
                    additionalSkillsTotal: 10
                },
                interviews: [
                    {
                        startTime: '5:55PM',
                        location: 'test 1'
                    },
                    {
                        startTime: '5:55PM',
                        location: 'test 2'
                    }
                ]
            };
            wrapper = setUp(props);
        });

        it('shows no dropdown for responses', () => {
            const leftColumnWrapper = wrapper.find('LeftColumn');
            expect(leftColumnWrapper).toMatchSnapshot();
        });

        it('shows a list of interviews and button to interviews', () => {
            const bottomWrapper = wrapper.find('Bottom');
            expect(bottomWrapper).toMatchSnapshot();
        });
    });
});
