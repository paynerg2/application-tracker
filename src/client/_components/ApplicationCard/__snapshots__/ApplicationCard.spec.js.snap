// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application Card Component when the component is not selected does not show additional buttons in the right column 1`] = `
<RightColumn>
  <div />
  <div>
    <div>
      test
    </div>
  </div>
  <div>
    5/5/1955
  </div>
</RightColumn>
`;

exports[`Application Card Component when the component is not selected returns to unselected state on mouse leave event 1`] = `
<RightColumn>
  <div />
  <div>
    <div>
      test
    </div>
  </div>
  <div>
    5/5/1955
  </div>
</RightColumn>
`;

exports[`Application Card Component when the component is selected shows additional buttons in the right column 1`] = `
<RightColumn>
  <div>
    <Link
      replace={false}
      to="/applications/test"
    >
      <Button>
        <ForwardRef />
      </Button>
    </Link>
    <Button
      onClick={[Function]}
    >
      <ForwardRef />
    </Button>
  </div>
  <div>
    <div>
      test
    </div>
  </div>
  <div>
    5/5/1955
  </div>
</RightColumn>
`;

exports[`Application Card Component when there are associated interviews shows a list of interviews and button to interviews 1`] = `
<Bottom>
  <div>
    <Connect(InterviewList)
      interviews={
        Array [
          Object {
            "location": "test 1",
            "startTime": "5:55PM",
          },
          Object {
            "location": "test 2",
            "startTime": "5:55PM",
          },
        ]
      }
    />
    <Button
      onClick={[Function]}
    >
      +
    </Button>
  </div>
  <div>
    <ProgressBar
      fill={10}
    />
    <ProgressBar
      fill={10}
    />
  </div>
</Bottom>
`;

exports[`Application Card Component when there are associated interviews shows no dropdown for responses 1`] = `
<LeftColumn>
  <div>
    <Title>
      test
    </Title>
    <div>
      test
    </div>
  </div>
  <Location>
    test
  </Location>
  <Bottom>
    <div>
      <Connect(InterviewList)
        interviews={
          Array [
            Object {
              "location": "test 1",
              "startTime": "5:55PM",
            },
            Object {
              "location": "test 2",
              "startTime": "5:55PM",
            },
          ]
        }
      />
      <Button
        onClick={[Function]}
      >
        +
      </Button>
    </div>
    <div>
      <ProgressBar
        fill={10}
      />
      <ProgressBar
        fill={10}
      />
    </div>
  </Bottom>
</LeftColumn>
`;

exports[`Application Card Component when there are no associated interviews does not show an interview list or button to add interviews 1`] = `
<Bottom>
  <div>
    <Dropdown
      onChange={[Function]}
      value="No Response"
    >
      <option
        value="No Response"
      >
        No Response
      </option>
      <option
        value="Rejected"
      >
        Rejected
      </option>
      <option
        value="Interview"
      >
        Add Interview
      </option>
    </Dropdown>
  </div>
  <div>
    <ProgressBar
      fill={10}
    />
    <ProgressBar
      fill={10}
    />
  </div>
</Bottom>
`;

exports[`Application Card Component when there are no associated interviews shows a dropdown for application response 1`] = `
<LeftColumn>
  <div>
    <Title>
      test
    </Title>
    <div>
      test
    </div>
  </div>
  <Location>
    test
  </Location>
  <Bottom>
    <div>
      <Dropdown
        onChange={[Function]}
        value="No Response"
      >
        <option
          value="No Response"
        >
          No Response
        </option>
        <option
          value="Rejected"
        >
          Rejected
        </option>
        <option
          value="Interview"
        >
          Add Interview
        </option>
      </Dropdown>
    </div>
    <div>
      <ProgressBar
        fill={10}
      />
      <ProgressBar
        fill={10}
      />
    </div>
  </Bottom>
</LeftColumn>
`;
